stablehlo/dialect/Base.h

```diff
@@ -51,6 +57,16 @@ bool isCompatibleForHloTypeInference(Type tp1, Type tp2);
 // sparsity.
 bool isCompatibleForHloTypeInference(TypeRange tp1, TypeRange tp2);
 
+// Infer single most specific return type from inputTypes with support for
+// bounds. (Size, bound) of each dimension of the return type will be inferred
+// from corresponding dimensions of every inputType by merging them. If
+// `concatenateDim` is set other than default -1, then that dimensions will be
+// concatenated into return type instead.
+LogicalResult inferMostSpecificType(Optional<Location> location,
```

Comment:
I have mixed emotions of using the same helper for both binary and concatenation, this helps avoid duplication but also a bit convoluted. Like, concat does all checks and handles unranked before this but check for bounds happen here.

Reply:
Thanks for the discussions! After deep thinking, I estimate that benefits is much less then the shortage to use a common function for both cases. So, I split it in the new commit. Let me answer the feedback first:
0. The `FailureOr<Type> mergeTypes()` interface above looks a clean kill. But in fact the concat op need both util methods `inferConcatenatedDimAndBound` for the concat_dim and `inferMergedDimAndBound` for the rest N-1 dims. Thus we have to pass in 2 functions instead of 1.
2 & 3: Due to my schedule, I put them into future "TODO(zhouxin) Refactor to better handle errors and return single type" in  `Base.cpp`
4. Fixed.

The biggest change in the newest commit is the full support of "mixed of unranked and ranked cases". I think this behavior issue is of top importance of this PR. See this [In the other comment above ](https://github.com/openxla/stablehlo/pull/389#discussion_r1008943148)

Reply:
To be honest, I liked the idea of sharing code between binary and concat, because this way we make sure that subtle logic like `<5x?xf32> op <*xf32>` is handled correctly. However given the schedule restrictions, I'm fine with merging as is and leaving follow-up to a ticket.

Reply:
Let's follow up in https://github.com/openxla/stablehlo/issues/409. Thanks for reviewing!